flashlight/fl/tensor/backend/jit/opt/backends/onednn/OneDnnOpFusion.cpp

```diff
@@ -0,0 +1,240 @@
+    case BinaryOp::Mul:
+      return dnnl::algorithm::binary_mul;
+    case BinaryOp::Div:
+      return dnnl::algorithm::binary_div;
+  }
+  throw std::runtime_error("Unsupported binary operation type");
+}
+
+bool isOpFusable(const BinaryOp op) {
+  switch (op) {
+    case BinaryOp::Add:
+    case BinaryOp::Mul:
+    case BinaryOp::Sub:
+    case BinaryOp::Div:
+      return true;
+  }
+  throw std::runtime_error("Unsupported binary operation type");
+}
+
+bool isNodeFusable(const Node* node) {
+  return node->isBinary() && isOpFusable(node->impl<BinaryNode>().op());
+}
+
+bool isFusionProfitable(const Node* node) {
+  // TODO Even if we have > 1 use, it might be possible & profitable to fuse,
+  // i.e., recomputation might be okay, think Halide.
+  return node->uses().size() <= 1;
+}
+
+bool shouldNodeBeFused(const Node* node) {
+  return isNodeFusable(node) && isFusionProfitable(node);
+}
+
+} // namespace
+
+Node* OneDnnOpFusion::rewriteFrom(Node* node) {
+  SearchState state{.searchRoot = node, .accumulatedBinopInfos = {}};
+  auto fusedNode = searchAndFuse(node, state);
+  node->replaceAllUsesWith(fusedNode);
+  return fusedNode;
+}
+
+Node* OneDnnOpFusion::searchAndFuse(Node* node, SearchState& state) {
+  // TODO for now we just skip shared input, need to think more.
+  if (visited_.find(node) != visited_.end() || !shouldNodeBeFused(node) ||
+      state.accumulatedBinopInfos.size() > kOneDnnMaxNumPostOps) {
+    return fuseNodes(node, state);
+  }
+  visited_.insert(node);
+
+  if (node->isBinary()) {
+    const auto& binaryNode = node->impl<BinaryNode>();
+    const auto lhs = binaryNode.lhs();
+    const auto rhs = binaryNode.rhs();
+    const auto op = binaryNode.op();
+    state.accumulatedBinopInfos.push_back({rewriteFrom(rhs), op});
+    return searchAndFuse(lhs, state);
+  } else {
+    // TODO support more fusion for more kinds of op (e.g., reduction)
+    throw std::runtime_error(
+        "[OneDnnOpFusion::rewriteFrom] If node should be fused, it must be binary node");
+  }
+}
+
+Node* OneDnnOpFusion::fuseNodes(Node* node, SearchState& state) {
+  for (const auto& input : node->inputs()) {
+    rewriteFrom(input);
+  }
+  // Nothing to fuse, it's one of the following:
+  // 1. node
+  //
+  // 2. node  ...
+  //      \  /
+  //    searchRoot
+  if (state.accumulatedBinopInfos.size() < 2) {
+    return state.searchRoot;
+  }
+
+  // In the following case `node` is `x1`
+  //
+  // x1  x2
+  //  \  /
+  //   op1  x3
+  //     \  /
+  //     op2
+  // becomes
+  // inputNodes: { x1, x2, x3 }
+  // algs:       { op1, op2 }
+  std::vector<Node*> inputNodes{node};
+  std::vector<dnnl::algorithm> algs;
+  for (int i = state.accumulatedBinopInfos.size() - 1; i >= 0; i--) {
+    const auto& info = state.accumulatedBinopInfos[i];
+    algs.push_back(binopToOneDnnAlg(info.op));
+    inputNodes.push_back(info.rhsNode);
+  }
+
+  // TODO refactor with common logic in OneDnnBackend
+  auto evalFunc = [algs = std::move(algs)](
```

Comment:
Great to have all of the goo we want to abstract away eventually in one place, nice.